### ISSUE TITLE
use MAP_JIT on MacOS

### ISSUE THREAD

**Issue by stsp** [dosemu2/dosemu2#2123]
According to https://github.com/dosemu2/dosemu2/issues/1005#issuecomment-1821069604
it may be interesting to share the
dosmem fd. It will also allow to clean
up the fdpp API a bit.
Unfortunately, the MacOS port currently
doesn't support this.

https://keith.github.io/xcode-man-pages/mmap.2.html
This page mentions MAP_JIT that seems
to allow PROT_WRITE|PROT_EXEC together
with MAP_SHARED.
@bartoldeman can you check out this
MAP_JIT thing?

**Comment by stsp** [comment]
Hi Bart, why not to check MAP_JIT on MacOS?
Should take a few minutes, not more...

**Comment by stsp** [comment]
I removed mapashm again, as it is
no longer compatible with native DPMI
(can't leverage the remote mapping hook).
So MAP_JIT needs to be tried instead.

**Comment by emendelson** [comment]
> Unfortunately, the MacOS port currently doesn't support this.

I've searched for the MacOS port everywhere I can think of, but can't find it. Can anyone point me in the right direction? Many thanks.

**Comment by stsp** [comment]
Maybe @bartoldeman knows better,
but afaik you have to build it there
from sources. I haven't set up any mac
build farm. :(

Среда, 2 июля 2025 г получено от emendelson:

**Comment by emendelson** [comment]
When I tried to build with `autogen.sh`, `default-configure`, and `make`, I got this error: `Please install i386-elf-binutils from https://github.com/nativeos/homebrew-i386-elf-toolchain`. I tried to do exactly that, following the instructions in the README.md file to the letter (how to tap the repository with brew), but I consistently got error messages, and the installation didn't work. 

I'll file a bug report with the `homebriew-i386-elf-toolchain` site in the hope of making it work. It would be terrific to have doesemu in macOS.

**Comment by stsp** [comment]
This is most likely simply unneeded.
Bart introduced this in 765f2e3f
(Sat Jul 1 13:02:45 2023), but later
82f9fe37 (Tue Sep 12 19:14:04 2023)
I started to support clang as as.

So I just removed this check.
I'll also see what can be done
with fdpp. While currently we
don't link fdpp with lld, there
are plenty of options. From
downloading the pre-built
fdppkernel.elf from CI, to
linking the unrelocatable
kernel.

**Comment by emendelson** [comment]
Will start a new issue on building under macOS:

https://github.com/dosemu2/dosemu2/issues/2491

**Comment by bartoldeman** [comment]
I haven't really touched dosemu for over a year, just regular life and work taking over.
I'll try to make it build again.

**Comment by stsp** [comment]
Wow, Bart is here!
You haven't touched dosemu
for over 2 years, to say the
least. :) Good to know everything
is allright.

**Comment by stuaxo** [comment]
@bartoldeman if it gets to the point it builds again, maybe it's worth adding to CI.

**Comment by stsp** [comment]
There are indeed a few github
action for this, like that one:
https://github.com/marketplace/actions/setup-osxcross-timmmm
Also there seems to be the
docker for off-line cross-compilation:
https://github.com/crazy-max/docker-osxcross
Does anyone want to try these
things out? :)

**Comment by stuaxo** [comment]
If I get it to build on work Mac, I'd give it a go...  I'll see what happens there first.

**Comment by stsp** [comment]
How to use that docker image?
I built it, but there are no further
instructions...

**Comment by stuaxo** [comment]
I can have a look at the docker file later.  In the meantime, I got as far as the libsearpc dependency, and then had trouble building that, so I'll look at that later too.

**Comment by stsp** [comment]
`--disable-searpc`

**Comment by stuaxo** [comment]
OK. configure completed.

```./configure --disable-fdpp --disable-searpc
...

Enabling plugins:

configure: WARNING: No video plugins available!

configure: WARNING: No audio plugins available!

configure: WARNING: fdpp not available!

configure: WARNING: searpc not available!

configure: WARNING: libacl not available!

configure: WARNING: libcap not available!
```

```sh
$ make
...
../../../src/plugin/plugin_config.hh:1:1: error: expected external declaration
    1 | -n
      | ^
../../../src/plugin/plugin_config.hh:1:2: error: unknown type name 'n'
    1 | -n
      |  ^
2 errors generated.
make[2]: *** [text.o] Error 1
make[1]: *** [base/video] Error 2
make: *** [default] Error 2
```

**Comment by stuaxo** [comment]
```
$ cat src/plugin/plugin_config.hh
-n
#include "../plugin/Xkmaps/config/plugin_config.h"
#include "../plugin/modemu/config/plugin_config.h"
```

**Comment by stsp** [comment]
You need at least sdl2!
Otherwise you build mostly
a non-functional thing.
Of course this won't solve
the compilation problem,
but at least configure should
do better.

**Comment by stsp** [comment]
> $ cat src/plugin/plugin_config.hh

In `scripts/mkpluginhooks` remove `-n`
after `echo` and try again.

**Comment by stuaxo** [comment]
That got further. 

I checked and "sh" doesn't seem to have -n on MacOS.  "dash", the minimal posix shell they have since Catalina does as well as bash and zsh.

```
$ make
...
ld: unknown options: -melf_i386
make[2]: *** [vesabios_pm.o] Error 1
make[1]: *** [base/dev/vga] Error 2
make: *** [default] Error 2
```

https://scriptingosx.com/2020/06/about-bash-zsh-sh-and-dash-in-macos-catalina-and-beyond/

**Comment by stsp** [comment]
What kind of ld is that?
Must be from cross binutils with
i386 elf support.

Пятница, 4 июля 2025 г получено от Stuart Axon:

**Comment by bartoldeman** [comment]
Yes, I got it to build again, but it only boots and crashes trying to load emufs, then can't read anything. Needs a few patches though.

"echo -n xxx" can be replaced by `printf %s xxx` except for "echo -n *", which is used in hdimage, which was my mistake 2 years ago, so Stas reversed it. `printf "%s " *` *may* do the trick (with a trailing space), or `echo * | tr -d '\n'`

You need to install `i386-elf-ld` for vesabios_pm.o. Configure will find it.
https://github.com/nativeos/homebrew-i386-elf-toolchain

there are a few other issues with linker flags (you need `-Wl,-undefined,dynamic_lookup` for dynamic plugins, which configure sets via `DOSEMU_LDFLAGS` but plugins don't use any more, on the other hand, `-Wl,-Bsymbolic -Wl,-soname=` is not understood). My build is quick and dirty, I'll get back to you later.

**Comment by bartoldeman** [comment]
Yes and indeed github actions supports a mac build, that'd would at least help avoiding accidentally breaking compilation.

**Comment by stsp** [comment]
Ci won't save the day, as once
something big is added again
(like was searpc or dj64 support),
then it will break and I won't have
an idea how to fix.
But it would still be good to have.
Can then be disabled, re-enabled and
so on.

Пятница, 4 июля 2025 г получено от Bart Oldeman:

**Comment by bartoldeman** [comment]
I realized I should still answer the original question and then continue in the new issue https://github.com/dosemu2/dosemu2/issues/2491

MAP_JIT won't help me here, as MAP_JIT|MAP_SHARED is rejected with "Invalid argument", it only works with ANON; and I can't get executable shared memory working on a Mac.

It's needed on Macs with ARM chips (Apple silicon) to create jit-able memory (really memory that can be toggled rx<->rw, rwx is simply not allowed there)

On Intel Macs (one of which I'm working with so far), rwx is allowed (only) on anon mappings without needing MAP_JIT, unless I use the hardened runtime (which is off by default).

this all goes back to the solution in
https://github.com/dosemu2/dosemu2/pull/2049#issuecomment-1630001169
but that same PR has another solution of simply not marking shared memory as executable unless native DPMI is used. I don't think native DPMI can work on MacOS anyway. Presently just getting the simulator back to work would be something.

**Comment by stuaxo** [comment]
> What kind of ld is that?
> Must be from cross binutils with
> i386 elf support.
> 
> Пятница, 4 июля 2025 г получено от Stuart Axon:
> […](#)

I think it's just the default one.

```
❯ which ld
/usr/bin/ld

░▒▓   …/dosemu2   devel !?   12:45 
❯ ld -v
@(#)PROGRAM:ld PROJECT:ld-1167.4.1
BUILD 20:00:10 Mar 12 2025
configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
will use ld-classic for: armv6 armv7 armv7s i386 armv6m armv7k armv7m armv7em
LTO support using: LLVM version 17.0.0 (static support for 29, runtime is 29)
TAPI support using: Apple TAPI version 17.0.0 (tapi-1700.0.3.4)
```
I tried:
`AS_LDFLAGS:=-arch i386 -platform_version macos 10.15 10.15`

To get this far:
```
/usr/bin/ld -arch i386 -platform_version macos 10.15 10.15 -static  -o vesabios_pm.o.elf _vesabios_pm.o
ld: warning: The i386 architecture is deprecated for macOS (remove from the Xcode build setting: ARCHS)
ld: warning: ignoring file _vesabios_pm.o, building for macOS-i386 but attempting to link with file built for unknown-unsupported file format ( 0x7F 0x45 0x4C 0x46 0x01 0x01 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 )
Undefined symbols for architecture i386:
  "start", referenced from:
     -u command line option
ld: symbol(s) not found for architecture i386
make[2]: *** [vesabios_pm.o] Error 1
make[1]: *** [base/dev/vga] Error 2
make: *** [default] Error 2
```

I tried another suggestion of using clang as linker `AS_LD:=/usr/bin/clang` though get the same output.

**Comment by stuaxo** [comment]
I got further specifying the entrypoint when building the vesa bios.

Current manual options in Makefile.conf:

```
XASFLAGS:=-x assembler -target i386-apple-darwin -c -
AS_LD:=/usr/bin/clang
AS_LDFLAGS:=-arch i386 -platform_version macos 10.15 10.15 -e _start
```

Gets to:

```
$ make
...
utilities.c:392:54: error: use of undeclared identifier 'memrchr'; did you mean 'memchr'?
  392 |                                 const char * slash = memrchr(res, '/', res_len);
      |                                                      ^~~~~~~
      |                                                      memchr
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/_string.h:74:3: note: 'memchr' declared here
   74 |                 memchr(const void *_LIBC_SIZE(__n) __s, int __c, size_t __n);
      |                 ^
1 error generated.
make[2]: *** [utilities.o] Error 1
make[1]: *** [base/misc] Error 2
make: *** [default] Error 2
```

EDIT:  It looks like `memrchr` isn't on MacOS

**Comment by bartoldeman** [comment]
@stuaxo I open coded `memrchr`.
Attached is a quick and dirty diff, I'll clean it up for a PR. I played the first level of the original Prince of Persia with sound using `$_cpuemu = (1)` (ie. using simulation) on an Intel Mac, was all good.

[dirtydiff.txt](https://github.com/user-attachments/files/21064568/dirtydiff.txt)

**Comment by stsp** [comment]
> ANON; and I can't get executable shared memory working on a Mac.

Hi Bart, please find the mapping
subsystem in a much better shape
than 2 years ago. :) We no longer
need executable shared mapping,
well unless for vm86. Native DPMI
is now remote, so at least in the
parent it doesn't need PROT_EXEC.
In #2492 I remove PROT_EXEC from
most mem_bases.

**Comment by stuaxo** [comment]
Thanks @bartoldeman I'll give it a go.  From what I saw implementing it is definitely an option on MacOS.

**Comment by stuaxo** [comment]
@bartoldeman I tried that branch, and needed the linker flags I added to get further (though I'm still not entirely sure they are correct).

then I got to a point where it looks like clangs assember and gnus are incompatible, with gnu `as` wanting `.globl` and clang wants `.global`
```
bios.S:616:9: error: non-local symbol required
 .globl LFN_short_name
        ^
bios.S:872:9: error: non-local symbol required
 .globl LFN_HELPER_OFF
        ^
bios.S:890:16: error: non-local symbol required
        .globl LFN_42_HELPER_OFF
               ^
bios.S:895:16: error: non-local symbol required
        .globl LFN_A6_HELPER_OFF
```

So maybe I should switch to building with gnu tools at this point.

**Comment by bartoldeman** [comment]
@stuaxo yes in my builds (my build PRs are now merged, no more patch needed) I'm still using https://github.com/nativeos/homebrew-i386-elf-toolchain

configure then detects:
```
XAS:='/usr/local/bin/i386-elf-as'
XASFLAGS:=
AS_LD:='/usr/local/bin/i386-elf-ld'
AS_LDFLAGS:=-melf_i386
```
we'll have then do some more changes to get your
```
XASFLAGS:=-x assembler -target i386-apple-darwin -c -
AS_LD:=/usr/bin/clang
AS_LDFLAGS:=-arch i386 -platform_version macos 10.15 10.15 -e _start
```
working

**Comment by stuaxo** [comment]
i386 elf toolchain failed to install, I got the error mentioned here:

https://github.com/nativeos/homebrew-i386-elf-toolchain/issues/29

**Comment by bartoldeman** [comment]
can you try to save the attached as `i386-elf-binutils.rb` and then run
`brew install --formula i386-elf-binutils.rb`
that'll try installing binutils 2.44, not sure if it works for you for it's worth a shot

[i386-elf-binutils.rb](https://github.com/user-attachments/files/21112464/i386-elf-binutils.rb.txt)

**Comment by stuaxo** [comment]
I had the same issue, though brew complained xcode is out of date 16.3 instead of 16.4 - I may need to wait until it filters through update system we have here.

**Comment by bartoldeman** [comment]
I found another way without `i386-elf-binutils`, by using the Apple clang assembler in combination with `ld.lld` from upstream LLVM (`brew install lld`)

then it compiles using:
```
gmake XASFLAGS="-x assembler -target i386-pc-linux-gnu -c -" AS_LD=ld.lld
```
the ` .globl LFN_short_name` error was misleading by the way; it understands both `.global` and `.globl` but for `-target i386-apple-darwin` it assumes symbols that start with the letter L are local. For `-target i386-pc-linux-gnu` it'll generates ELF objects like GAS without that restriction.

**Comment by stuaxo** [comment]
That is a sneaky error.   When I tried `as` from `clang` I tried setting up a GLOBAL define, the thing it complained about then was that the GLOBAL bit had to be after the thing being defined, not before it.

Do I run that command from the project root ?

**Comment by bartoldeman** [comment]
yes, in actual fact configure already finds `AS=ld.lld` if you have it, there is even logic in there for
```
XASFLAGS="-x assembler -target i686-unknown-linux-gnu -c -"
```
(does the same thing as i386-pc-linux-gnu), it's just not triggered on MacOS yet.

**Comment by stsp** [comment]
So what's the state of MacOS support?
I guess this all was solved, and
MAP_JIT is not needed?

**Comment by bartoldeman** [comment]
I'll have to retry on the Intel Mac (where MAP_JIT was indeed not needed) but don't have an Apple Silicon (Arm) Mac to try it on. I suspect some of the recent changes with larger page sizes and softmmu would have helped on Arm Mac as well (since it also uses 16k pages).
